service/api_logic/auth_strategy.py

```diff
@@ -0,0 +1,89 @@
+from abc import ABC, abstractmethod
+from typing import Dict
+import bcrypt
+import requests
+from flask import current_app, request
+from itsdangerous import URLSafeTimedSerializer
+from oauthlib.oauth2 import WebApplicationClient
+
+from database.models import User
+from dto.api_input import InputUserLoginDTO, InputUserByGoogleDTO
+from dto.api_output import OutputLogin
+from exept.exeptions import IncorrectUsernameOrEmailError, IncorrectPasswordError
+from typing import Generic, TypeVar
+
+from logger.logger import Logger
+
+T = TypeVar("T")
+
+class AuthStrategy(ABC, Generic[T]):
+    def __init__(self, user_dal, serializer):
+        self._user_dal = user_dal
+        self._logger = Logger("logger", "all.log").logger
+        self._serializer = serializer
+
+    @abstractmethod
+    def authenticate(self, credentials: T):
+        pass
+
+
+class AuthContext:
+    def __init__(self, strategy: AuthStrategy):
+        self._strategy = strategy
+
+    async def execute_login(self, credentials: T):
+        xuy = await self._strategy.authenticate(credentials)
+        return xuy
+
+
+class SimpleAuthStrategy(AuthStrategy[InputUserLoginDTO]):
+    async def authenticate(self, credentials: InputUserLoginDTO):
```

Comment:
When you change the DTO on API level to be general, you won't need `InputUserLoginDTO` and `InputUserByGoogleDTO` since they will basically merge